Naughts_and_Crosses.py

```diff
@@ -1,122 +1,237 @@
+                    == self.current_player.symbol for j in range(self.board.board_size)
+                    ):
+                    print(self.current_player.name + " wins! Well done")
+                    self.running = False
+                    return True
+
+    def end_game(self):  #Tells then to bugger off
+        print("Game's over, go home now... Unless...")
+
+    def _player1_assignment(self): #Asks the user to say if player 1 is a human or a robot
+        self.player1_type = (input("Is player 1 a human or a robot?  (h for Human or a for Ai)"  )).lower()
+        if self.player1_type != 'h' and self.player1_type != 'a':
+            self._player1_assignment()
+
+    def _player2_assignment(self): #Asks the user to say if player 2 is a human or a robot
+        self.player2_type = (input("Is player 2 a human or a robot? (h for Human or a for Ai)  ")).lower()
+        if self.player2_type != 'h' and self.player2_type != 'a':
+            self._player2_assignment()
+
+    def _generate_player1(self): #Depending on the choice the user made, generates player1 as a Human or Ai
+        if self.player1_type == 'h':  #then assigns the player their range of choices based upon the board size
+            self.p1 = HumanPlayer()
+        else:
+            self.p1 = AiPlayer()
+        self.p1.choice_range = self.get_board_size()
+
+    def _generate_player1_symbol(self):  #Asks the user which symbol player 1 wants to be
+        player1_symbol = input('What symbol would you like player 1 to be? (X or O)  ')
+        if player1_symbol == 'X':
+            self.p1.symbol = 'X'
+        elif player1_symbol == 'O':
+            self.p1.symbol = 'O'
+        else:
+            _generate_player1_symbol()
 
+    def _generate_player2(self):  #Depding on the choice, generates play2 as a Human or AI and assigns the player
+        if self.player2_type == 'h':  #their range of choices based on the board_size
+            self.p2 = HumanPlayer()
+        else:
+            self.p2 = AiPlayer()
+        self.p2.choice_range = self.get_board_size()
+
+    def _generate_player2_symbol(self): #Depending on what Player1's symbol is, assigns Player2 their symbol
+        if self.p1.symbol == 'X':
+            self.p2.symbol = 'O'
+        elif self.p1.symbol == 'O':
+            self.p2.symbol = 'X'
+        
+    def _generate_players(self): #Running order to generate the players
+        self._player1_assignment()
+        self._generate_player1()
+        self._generate_player1_symbol()
+        self._player2_assignment()
+        self._generate_player2()
+        self._generate_player2_symbol()
+
+    def get_board_size(self): #Used to get the board size in the board object in order to pass it to the player objects
+    	return self.board.board_size
+
+class HumanPlayer():
+    def __init__(self): #Asks the person their name to be used in all further communications
+        self.name = input('What is your name?  ')
+        self.player_type = 'H'
```

Comment:
Define this outside the function, like:

```python
class HumanPlayer():
    player_type = 'H'

    def __init__(self):
        blah
````

Again, define the string variables as globals.